llvm/test/Transforms/InstCombine/free-inversion.ll

```diff
@@ -0,0 +1,521 @@
+; CHECK-NEXT:    [[NNA:%.*]] = xor i8 [[NA:%.*]], -1
+; CHECK-NEXT:    [[A:%.*]] = sub i8 [[NNA]], [[AA:%.*]]
+; CHECK-NEXT:    [[AB:%.*]] = call i8 @llvm.smax.i8(i8 [[A]], i8 [[B]])
+; CHECK-NEXT:    [[NOT_AB:%.*]] = xor i8 [[AB]], -1
+; CHECK-NEXT:    ret i8 [[NOT_AB]]
+;
+  %nx = xor i8 %x, -1
+  %yy = xor i8 %y, 123
+  %b = select i1 %c, i8 %nx, i8 %yy
+  %nna = xor i8 %na, -1
+  %a = sub i8 %nna, %aa
+  %ab = call i8 @llvm.smax.i8(i8 %a, i8 %b)
+  %not_ab = xor i8 %ab, -1
+  ret i8 %not_ab
+}
+
+define i8 @smax_1_fail(i8 %aa, i8 %na, i1 %c, i8 %x, i8 %y) {
+; CHECK-LABEL: @smax_1_fail(
+; CHECK-NEXT:    [[NX:%.*]] = xor i8 [[X:%.*]], -1
+; CHECK-NEXT:    [[YY:%.*]] = xor i8 [[Y:%.*]], 123
+; CHECK-NEXT:    call void @use.i8(i8 [[YY]])
+; CHECK-NEXT:    [[B:%.*]] = select i1 [[C:%.*]], i8 [[NX]], i8 [[YY]]
+; CHECK-NEXT:    [[NNA:%.*]] = xor i8 [[NA:%.*]], -1
+; CHECK-NEXT:    [[A:%.*]] = sub i8 [[NNA]], [[AA:%.*]]
+; CHECK-NEXT:    [[AB:%.*]] = call i8 @llvm.smax.i8(i8 [[A]], i8 [[B]])
+; CHECK-NEXT:    [[NOT_AB:%.*]] = xor i8 [[AB]], -1
+; CHECK-NEXT:    ret i8 [[NOT_AB]]
+;
+  %nx = xor i8 %x, -1
+  %yy = xor i8 %y, 123
+  call void @use.i8(i8 %yy)
+  %b = select i1 %c, i8 %nx, i8 %yy
+  %nna = xor i8 %na, -1
+  %a = sub i8 %nna, %aa
+  %ab = call i8 @llvm.smax.i8(i8 %a, i8 %b)
+  %not_ab = xor i8 %ab, -1
+  ret i8 %not_ab
+}
+
+define i8 @umax_1(i8 %na, i1 %c, i8 %x, i8 %y) {
+; CHECK-LABEL: @umax_1(
+; CHECK-NEXT:    [[NX:%.*]] = xor i8 [[X:%.*]], -1
+; CHECK-NEXT:    [[YY:%.*]] = xor i8 [[Y:%.*]], 123
+; CHECK-NEXT:    [[B:%.*]] = select i1 [[C:%.*]], i8 [[NX]], i8 [[YY]]
+; CHECK-NEXT:    [[BA:%.*]] = call i8 @llvm.umax.i8(i8 [[B]], i8 85)
+; CHECK-NEXT:    [[NOT_BA:%.*]] = xor i8 [[BA]], -1
+; CHECK-NEXT:    ret i8 [[NOT_BA]]
+;
+  %nx = xor i8 %x, -1
+  %yy = xor i8 %y, 123
+  %b = select i1 %c, i8 %nx, i8 %yy
+  %ba = call i8 @llvm.umax.i8(i8 %b, i8 85)
+  %not_ba = xor i8 %ba, -1
+  ret i8 %not_ba
+}
+
+define i8 @umax_1_fail(i8 %na, i1 %c, i8 %x, i8 %y) {
+; CHECK-LABEL: @umax_1_fail(
+; CHECK-NEXT:    [[NX:%.*]] = xor i8 [[X:%.*]], -1
+; CHECK-NEXT:    [[YY:%.*]] = xor i8 [[Y:%.*]], 123
+; CHECK-NEXT:    [[B:%.*]] = select i1 [[C:%.*]], i8 [[NX]], i8 [[YY]]
+; CHECK-NEXT:    call void @use.i8(i8 [[B]])
+; CHECK-NEXT:    [[BA:%.*]] = call i8 @llvm.umax.i8(i8 [[B]], i8 85)
+; CHECK-NEXT:    [[NOT_BA:%.*]] = xor i8 [[BA]], -1
+; CHECK-NEXT:    ret i8 [[NOT_BA]]
+;
+  %nx = xor i8 %x, -1
+  %yy = xor i8 %y, 123
+  %b = select i1 %c, i8 %nx, i8 %yy
+  call void @use.i8(i8 %b)
+  %ba = call i8 @llvm.umax.i8(i8 %b, i8 85)
+  %not_ba = xor i8 %ba, -1
+  ret i8 %not_ba
+}
+
+define i8 @sub_both_freely_invertable_always(i8 %x, i8 %y) {
+; CHECK-LABEL: @sub_both_freely_invertable_always(
+; CHECK-NEXT:    [[XX:%.*]] = xor i8 [[X:%.*]], 123
+; CHECK-NEXT:    [[TMP1:%.*]] = xor i8 [[Y:%.*]], -46
+; CHECK-NEXT:    [[YY_NEG:%.*]] = add i8 [[TMP1]], 1
+; CHECK-NEXT:    [[R:%.*]] = add i8 [[YY_NEG]], [[XX]]
```

Comment:
Pre-existing issue: We're adding an extra instruction here. It looks like this bit in InstCombineNegator is responsible: https://github.com/llvm/llvm-project/blob/8748b3979ce414025825631dd77c27c58a4ed04b/llvm/lib/Transforms/InstCombine/InstCombineNegator.cpp#L445-L455 It's missing a IsTrulyNegation check. Will we run into an infinite loop here once we add it?

Reply:
Yeah, you're right. Ill post patch for `IsTrulyNegatable` and add option for `MustConsumeInst` in `getFreelyInverted`.

Reply:
See: https://github.com/llvm/llvm-project/pull/72767, will repost w/ `MustConsumeInst` flag once thats in.

Reply:
Okay, update so that we ensure we save an inst.